texar/evals/bleu_tool.py

```diff
@@ -0,0 +1,217 @@
+
+    for i in range(max_order):
+        if possible_matches_by_order[i] > 0:
+            precisions[i] = matches_by_order[i] / possible_matches_by_order[i]
+            if matches_by_order[i] > 0:
+                precisions[i] = (matches_by_order[i] /
+                                 possible_matches_by_order[i])
+            else:
+                smooth *= 2
+                precisions[i] = 1.0 / (smooth * possible_matches_by_order[i])
+        else:
+            precisions[i] = 0.0
+
+    if max(precisions) > 0:
+        p_log_sum = sum(math.log(p) for p in precisions if p)
+        geo_mean = math.exp(p_log_sum / max_order)
+
+    if use_bp:
+        ratio = translation_length / reference_length
+        if ratio <= 0:
+            bp = 0
+        elif ratio < 1.0:
+            bp = math.exp(1 - 1.0 / ratio)
+        else:
+            bp = 1.0
+
+    bleu = geo_mean * bp
+
+    return np.float32(bleu) * 100
+
+
+class UnicodeRegex:
+    """Ad-hoc hack to recognize all punctuation and symbols."""
+
+    def __init__(self):
+        punctuation = self.property_chars("P")
+        self.nondigit_punct_re = re.compile(r"([^\d])([" + punctuation + r"])")
+        self.punct_nondigit_re = re.compile(r"([" + punctuation + r"])([^\d])")
+        self.symbol_re = re.compile("([" + self.property_chars("S") + "])")
+
+    @staticmethod
+    def property_chars(prefix):
+        return "".join(
+            chr(x)
+            for x in range(sys.maxunicode)
+            if unicodedata.category(chr(x)).startswith(prefix)
+        )
+
+
+uregex = UnicodeRegex()
+
+
+def bleu_tokenize(string: str) -> List[str]:
+    r"""Tokenize a string following the official BLEU implementation.
+
+    See
+    `https://github.com/moses-smt/mosesdecoder/blob/master/scripts/generic/mteval-v14.pl#L954-L983`.
+    In our case, the input string is expected to be just one line
+    and no HTML entities de-escaping is needed.
+    So we just tokenize on punctuation and symbols,
+    except when a punctuation is preceded and followed by a digit
+    (e.g. a comma/dot as a thousand/decimal separator).
+
+    Note that a number (e.g. a year) followed by a dot at the end of sentence
+    is NOT tokenized,
+    i.e. the dot stays with the number because `s/(\p{P})(\P{N})/ $1 $2/g`
+    does not match this case (unless we add a space after each sentence).
+    However, this error is already in the original `mteval-v14.pl`
+    and we want to be consistent with it.
+
+    Args:
+        string: the input string
+
+    Returns:
+        a list of tokens
+    """
+    string = uregex.nondigit_punct_re.sub(r"\1 \2 ", string)
+    string = uregex.punct_nondigit_re.sub(r" \1 \2", string)
+    string = uregex.symbol_re.sub(r" \1 ", string)
+    return string.split()
+
+
+def bleu_wrapper(ref_filename: str,
+                 hyp_filename: str,
+                 case_sensitive: bool = False) -> float:
+    r"""Compute BLEU for two files (reference and hypothesis translation).
```

Comment:
Let's call it `file_bleu_transformer`, indicating it's for files (on par w/ `sentence_*`, `corpus_*`, etc), and it's the `*_transformer` version.

Even better, can we have a general `file_bleu` with an option for users to choose the BLEU version (if the file reading and preprocessing can be shared)?

Reply:
Either way, the docstring should be more detailed, especially indicating which BLEU version is used.